Add scalar autograd tests for functions requiring 'special' Variables…

gchanan opened the pull request:
… on LHS.

gchanan committed:
Add scalar autograd tests for functions requiring 'special' Variables on LHS.

gchanan committed:
Properly fill in make_non_contiguous data for sizes that can't be mad… (#4951)

* Properly fill in make_non_contiguous data for sizes that can't be made contiguous.

* Use clone instead of copy.

gchanan committed:
Add index_* tests.

gchanan committed:
Fix flake8.

gchanan committed:
Use normal for clamp rather than uniform.

gchanan committed:
Add tests for gather, scatter, scatter_add.

gchanan committed:
Make sure masked_select doesn't get all zeros.

gchanan committed:
Fix and test backward for mv, ger with scalars.

gchanan committed:
Fix addmv.

colesbury commented on torch/autograd/gradcheck.py:
```diff
@@ -203,6 +203,10 @@ def fn(input):
                 continue
             if not i.grad.data.eq(0).all():
                 return fail_test('backward not multiplied by grad_output')
+            if type(i.grad) != type(i):
```
Use `i.grad.type()` and `i.type()` so that it will work with variables

gchanan committed:
Use grad.type() instead of type(grad).

gchanan committed:
Fix addr.

There are a couple of hacks here:
1) We need to squeeze the backward result because of implicit broadcast of the arguments to match behavior of ger.
2) The broadcast_dims code doesn't work for scalars; I added support for adding '.scalar' onto the end of the broadcast
   specification, but really this should just be a native function with _out support.

gchanan committed:
Don't allow scalars in torch.dot for Variables.

There is no dot_out, so the lack of _out isn't an issue.

colesbury commented:
I don't think we should allow scalars in `addr` unless we're going to add support for N-dimensional tensors.

gchanan commented:
That same logic should apply to `ger` and should be easy to do assuming we are willing to get rid of the `out` variants -- do you want to do that?

gchanan commented:
ok, I'm going to back out the contentious stuff; anything we do would probably be done as a native function, where we would want `out` support; I'll add that in a separate PR and we'll decide what to do after that.

gchanan committed:
Revert "Don't allow scalars in torch.dot for Variables."

This reverts commit 76c521eba8c1fb533e164f121075230209d52927.

gchanan committed:
Revert "Fix addr."

This reverts commit afe04a0078394f94645e10cec53626f582cbc55c.

gchanan committed:
Revert "Fix addmv."

This reverts commit 550c7ac71b3b832a3b74a809fec9ce5f5e554909.

gchanan committed:
Revert "Use grad.type() instead of type(grad)."

This reverts commit ddcb5a424ed004fa2ee238a50177573e6d4a1b89.

gchanan committed:
Revert "Fix and test backward for mv, ger with scalars."

This reverts commit 10b0ecad48d987774c41184ffaf11742322926ab.

gchanan commented:
@pytorchbot retest this please.

gchanan commented:
Onnx failure looks unrelated.
